src/components/postgresConnect.ts

```diff
@@ -1,45 +1,210 @@
+import { Client, Pool, QueryResult } from 'pg'
+import {
+  ActionInTemplate,
+  ActionQueue,
+  ActionPlugin,
+  File,
+  ActionInTemplateGetPayload,
+  ActionQueueExecutePayload,
+  ActionQueueGetPayload,
+  ActionQueuePayload,
+  ActionPluginDeletePayload,
+  ActionPluginPayload,
+  FilePayload,
+  FileGetPayload,
+  TriggerQueueUpdatePayload,
+} from '../types'
 
-const { Client } = require('pg')
+class PostgresDB {
+  private static _instance: PostgresDB
+  private pool: Pool
 
-// Postgres Database listeners for Triggers/Actions
-export const pgClient = new Client(config.pg_database_connection)
+  constructor() {
+    this.pool = new Pool(config.pg_database_connection)
 
-pgClient.connect()
-console.log('Connecting to Postgres...')
+    // the pool will emit an error on behalf of any idle pools
+    // it contains if a backend error or network partition happens
+    this.pool.on('error', (err) => {
+      console.error('Unexpected error on idle pool', err)
+      process.exit(-1)
+    })
 
-// Load action plugins
-const actionLibrary: ActionLibrary = {}
-const actionSchedule: any[] = []
+    this.startListener()
+  }
 
-export const loadActionPlugins = async () => {
-  // Scan plugins folder and update Database
-  await registerPlugins(pgClient)
-  // Load Action functions into global scope
-  await loadActions(pgClient, actionLibrary)
-  // Schedule future actions and execute overdue ones
-  await loadScheduledActions(pgClient, actionLibrary, actionSchedule)
-}
+  private startListener = async () => {
+    const listener = new Client(config.pg_database_connection)
+    listener.connect()
+    listener.query('LISTEN trigger_notifications')
+    listener.query('LISTEN action_notifications')
+    listener.on('notification', ({ channel, payload }) => {
+      if (!payload) {
+        console.log(`Notification ${channel} received with no payload!`)
+        return
+      }
+      switch (channel) {
+        case 'trigger_notifications':
+          processTrigger(JSON.parse(payload))
+          break
+        case 'action_notifications':
+          executeAction(JSON.parse(payload), actionLibrary)
+          break
+      }
+    })
+  }
+
+  private getValuesPlaceholders = (object: { [key: string]: any }) =>
+    Object.keys(object).map((key, index) => `$${index + 1}`)
+
+  public static get Instance() {
+    return this._instance || (this._instance = new this())
+  }
+
+  public query = async (text: string, payload: any[]): Promise<QueryResult> => {
+    const client = await this.pool.connect()
+    try {
+      return await client.query(text, payload)
+    } finally {
+      // Make sure to release the client before any error handling,
+      // just in case the error handling itself throws an error.
+      client.release()
+    }
+  }
+
+  public addActionQueue = async (action: ActionQueuePayload) => {
+    try {
+      const result = await this.query(
+        `INSERT into action_queue (${Object.keys(
```

Comment:
I guess the reason you're building this query string yourself rather than passing substitution parameters is because the substitutions are arrays rather than strings?

It's highly recommended to not do this, as the built-in parameter parsing is much more robust against SQL injection. I think you should be able to pass arrays as parameters -- node-pg will convert them to Postgres arrays. (See the "parameterized queries" section here: https://node-postgres.com/features/queries)

You might need to use this module if you can't do it with the current tools: https://www.npmjs.com/package/pg-format

Reply:
The keys and objects passed have to be in the format expected by `ActionQueuePayload`. I tried to see this package `pg-format` before but it wasn't clear to me how to use in the same way that we would like to: with keys and values that are variable, not always the full structure. I need some more input on this one to continue.

Reply:
Okay, I think this works adequately:
```
const query = `INSERT into action_queue (${Object.keys(action
      )}, time_queued) VALUES (${this.getValuesPlaceholders(action)}, CURRENT_TIMESTAMP)`
const result = await this.query(query, Object.values(action))
```

Reply:
Nice. Thanks for looking into this with more detail. Is always better to be sure the core is making safe stuff.

Reply:
Ok. I have updated in every INSERT query that was using the dynamic keys and placeholders to build the query. :)